### PR TITLE
Center the sub-panels (Quest, Character, Spell and Inventory)

### PR DESCRIPTION
**Pull request by 4aron5un**
According to screen resolution to center the sub-panels

[1440x900]
1440x900_Char_Inv
1440x900_Quest_Spell

[800x600]
800x600_Char_Inv
800x600_Quest_Spell

[640x480]
640x480_Char_Inv
640x480_Quest_Spell

### CODE REVIEW

**Comment by 4aron5un** [comment]
Centered_Panels

**Comment by AJenbo** [comment]
How does this look in the more common case of 853x480? The benifit of having them to the side is that you can more reasonably play while they are open. Though this might not be ideal for touch controls.

**Comment by 4aron5un** [comment]
Thanks.
I got your point about it, and how about this?

Centered_Panels_Side

Attached 853x480 for your reference

[Current]
853x480_Char_Inv_Current
[Centered]
853x480_Char_Inv

[Current]
853x480_Quest_Spell_Current
[Centered]
853x480_Quest_Spell

**Comment by AJenbo** [comment]
> I got your point about it, and how about this?
> 
> Centered_Panels_Side

This looks good :+1:

**Comment by 4aron5un** [comment]
Just to share my experience...:)

For desktop user with high resolution and large monitor, current panel is far for me.

So I suggest moving them to the visual focus area.

But for mobile phone players, the touch area is very important.

Maybe we can try this "vertical centered" version, I will commit later...:)

1920x1080_Char_Inv_Current

1920x1080_Char_Inv_Centered

Visual_Focus

**Comment by 4aron5un** [comment]
Vertical centered

[640x480]
640x480_Vertical_Centered


[853x480]
853x480_Vertical_Centered


[1440x900]
1440x900_Vertical_Centered


[1920x1080]
1920x1080_Vertical_Centered

**Comment by AJenbo** [comment]
:+1: eventually we can add an option that allows the user to pick where they should be positioned. Doing so would also get rid of the preprocessor defines which would help to improve the code.

**Comment by glebm** [comment]
This is a neat refactoring! Perhaps the refactoring can be merged separately, and then a follow-up PR adjusting the positions?

**Comment by julealgon** [comment]
What about allowing to move the windows in-game, and then persist their position in the save, or in the ini (easier, and would then be global)?

As an alternative, just allowing to set their positions in the ini would be a good start.

**Comment by FitzRoyX** [comment]
Yeah, the cornered panels were not ideal for 720p and up. Too much mouse travel. Bringing them down is a welcome change. I think a clamped solution where high res can enjoy the low mouse travel of "floating" panels without being obstructive to ultra-low resolutions is best.

**Comment by 4aron5un** [comment]
Thank you all for the great input :blush:

Please check my TODOs below:
- Separate this change into a macro definition, just like what NONET/USE_SDL1 did
- Disable this feature by default, developer can enable this feature manually
- After we have enough verification, maybe we can consider to enable it by default
- Provide several options in the .ini file, such like (0: Align to the Corner, 1: Vertically Centered, 2: Fully Centered, ...)
- And the next phase, maybe we can implement the in-game adjustment and persist the last position

**Comment by julealgon** [comment]
> Thank you all for the great input 😊
> 
> Please check my TODOs below:
> 
> * Separate this change into a macro definition, just like what NONET/USE_SDL1 did

Why? I think we should avoid macros as much as possible for new code. This is a bad idea imho.

> * Disable this feature by default, developer can enable this feature manually

If the settings for changing the positions are in the .ini, there should be no need for a "master toggle" of the feature.

> * After we have enough verification, maybe we can consider to enable it by default

See above. I disagree with this.

> * Provide several options in the .ini file, such like (0: Align to the Corner, 1: Vertically Centered, 2: Fully Centered, ...)

That's an option, but it might be simpler to just allow one to set the corner(s) position in pixels instead. This would even allow people to swap the positions of the panels entirely (say, have the inventory on the left, and the character screen on the right, etc) or even have the panels one on top of the other in high resolutions.

> * And the next phase, maybe we can implement the in-game adjustment and persist the last position

Yeah I think being able to move the windows inside the game could definitely be a separate, incremental effort from this.

**Comment by AJenbo** [comment]
>     * Separate this change into a macro definition, just like what NONET/USE_SDL1 did
>     * Disable this feature by default, developer can enable this feature manually
>     * After we have enough verification, maybe we can consider to enable it by default

I don't think any of that is needed.

>     * Provide several options in the .ini file, such like (0: Align to the Corner, 1: Vertically Centered, 2: Fully Centered, ...)

Sounds good as a follow up.
 
>     * And the next phase, maybe we can implement the in-game adjustment and persist the last position

I think this is overkill, but I'm open to be proven wrong.

**Comment by 4aron5un** [comment]
> > Thank you all for the great input blush
> > Please check my TODOs below:
> > 
> > * Separate this change into a macro definition, just like what NONET/USE_SDL1 did
> 
> Why? I think we should avoid macros as much as possible for new code. This is a bad idea imho.
> 
> > * Disable this feature by default, developer can enable this feature manually
> 
> If the settings for changing the positions are in the .ini, there should be no need for a "master toggle" of the feature.
> 
> > * After we have enough verification, maybe we can consider to enable it by default
> 
> See above. I disagree with this.

OK, got it. Just worried about my change will impact other function in the formal build.
> > * Provide several options in the .ini file, such like (0: Align to the Corner, 1: Vertically Centered, 2: Fully Centered, ...)
> 
> That's an option, but it might be simpler to just allow one to set the corner(s) position in pixels instead. This would even allow people to swap the positions of the panels entirely (say, have the inventory on the left, and the character screen on the right, etc) or even have the panels one on top of the other in high resolutions.
> 
> > * And the next phase, maybe we can implement the in-game adjustment and persist the last position
> 
> Yeah I think being able to move the windows inside the game could definitely be a separate, incremental effort from this.

Thanks for the comment. Let's have an ini version first. :smile:

**Comment by 4aron5un** [comment]
> > ```
> > * Separate this change into a macro definition, just like what NONET/USE_SDL1 did
> > * Disable this feature by default, developer can enable this feature manually
> > * After we have enough verification, maybe we can consider to enable it by default
> > ```
> 
> I don't think any of that is needed.
> 
> > ```
> > * Provide several options in the .ini file, such like (0: Align to the Corner, 1: Vertically Centered, 2: Fully Centered, ...)
> > ```
> 
> Sounds good as a follow up.
> 
Got it, I will implement options in .ini file.
> > ```
> > * And the next phase, maybe we can implement the in-game adjustment and persist the last position
> > ```
> 
> I think this is overkill, but I'm open to be proven wrong.

**Comment by julealgon** [comment]
> OK, got it. Just worried about my change will impact other function in the formal build.

If it introduces any issues, we fix them. 

The toggles we currently have in ini are not for this purpose (fear of breaking code/being easy to revert) but for toggling between 2 desirable behavior states for the user.

File: Source/control.h [hunk 1]
```diff
@@ -27,8 +27,10 @@ namespace devilution {
 
 #define SPANEL_WIDTH 320
 #define SPANEL_HEIGHT 352
-#define RIGHT_PANEL (gnScreenWidth - SPANEL_WIDTH)
-#define RIGHT_PANEL_X RIGHT_PANEL
+#define LEFT_PANEL_X 0
+#define LEFT_PANEL_Y ((gnScreenHeight - PANEL_HEIGHT - SPANEL_HEIGHT) / 2)
+#define RIGHT_PANEL_X (gnScreenWidth - SPANEL_WIDTH)
+#define RIGHT_PANEL_Y LEFT_PANEL_Y
```
Comment (AJenbo) [inline Source/control.h:33]: When changing this to be configurable these should probably be replaced with a function that returns a point when given an enum that represents one of the 4 panels.

`Point GetPanelPosition(UiPanels panel);`

File: Source/diablo.cpp [hunk 2]
```diff
@@ -327,17 +327,17 @@ void LeftMouseDown(int wParam)
 
 	if (MousePosition.y < PANEL_TOP || MousePosition.x < PANEL_LEFT || MousePosition.x >= PANEL_LEFT + PANEL_WIDTH) {
 		if (!gmenu_is_active() && !TryIconCurs()) {
-			if (QuestLogIsOpen && MousePosition.x > 32 && MousePosition.x < 288 && MousePosition.y > 32 && MousePosition.y < 308) {
+			if (QuestLogIsOpen && MousePosition.x > LEFT_PANEL_X + 32 && MousePosition.x < LEFT_PANEL_X + 288 && MousePosition.y > LEFT_PANEL_Y + 32 && MousePosition.y < LEFT_PANEL_Y + 308) {
```
Comment (AJenbo) [inline Source/diablo.cpp:330]: Stuff like this should end up some thing like this:

```cpp
for (auto panelType : UiPanels) {
    if (!IsPanelOpen(panelType)
        continue;
        
    Rectangle panel { GetPanelPosition(panelType), { 320, 352 } };
    if (panel.Contains(MousePosition)) {
        return true;
    }
}

return false;
```

**Review by AJenbo — APPROVED** [review]

File: Source/control.h (hunk 1 above)
Comment (4aron5un) [reply Source/control.h:33]: I will do the refactoring for this.

File: Source/diablo.cpp (hunk 2 above)
Comment (4aron5un) [reply Source/diablo.cpp:330]: Got it.

This will become a function such like IsCursorOnPanel().

**Pull request merged**